### PR TITLE
fix(migration): Fix duplicate Null Users issue

### PR DESCRIPTION
**Pull request by justin-tahara**
## Description

This PR fixes a migration failure in `e7f8a9b0c1d2_create_anonymous_user` for tenants with legacy `notification` rows where `user_id` is `NULL`.

The root cause is an interaction between old data and a newer uniqueness rule. We added a unique index on `notification` over `(user_id, notif_type, COALESCE(additional_data, '{}'::jsonb))` to support notification deduplication. PostgreSQL still allows multiple rows with the same `notif_type` and `additional_data` when `user_id` is `NULL`, because `NULL` values do not conflict in a unique index. Later, the anonymous-user migration tries to reassign all remaining `NULL`-owned rows to the fixed anonymous user UUID. When duplicate `NULL`-owned notifications are updated to the same non-null `user_id`, they become true duplicates and the update fails with a unique-constraint violation. Because the migration runs inside a single Alembic transaction, that one failure aborts the transaction and prevents the revision from being recorded.

This PR fixes the migration in two ways. First, before updating `notification.user_id`, it deduplicates `NULL`-owned notifications and removes any rows that would also conflict with an already-existing anonymous-user notification. Second, it wraps each table-level migration step in a nested transaction/savepoint so a handled failure does not poison the entire Alembic transaction. A regression test was added to cover the failing case: two identical `NULL`-owned notifications existing before `e7f8a9b0c1d2`, and the migration completing successfully with a single anonymous-owned notification remaining.

I explicitly needed to run the following command in order to unblock them: 
```
BEGIN;

SET search_path TO public;

-- Preview duplicate null-owned notifications that will collide on migration
SELECT
  notif_type,
  COALESCE(additional_data, '{}'::jsonb) AS additional_data,
  COUNT(*) AS cnt,
  ARRAY_AGG(id ORDER BY first_shown DESC, last_shown DESC, id DESC) AS ids
FROM notification
WHERE user_id IS NULL
GROUP BY notif_type, COALESCE(additional_data, '{}'::jsonb)
HAVING COUNT(*) > 1
ORDER BY cnt DESC, notif_type;

-- Delete all but the newest row in each duplicate NULL-owned notification group
WITH ranked AS (
  SELECT
    id,
    ROW_NUMBER() OVER (
      PARTITION BY notif_type, COALESCE(additional_data, '{}'::jsonb)
      ORDER BY first_shown DESC, last_shown DESC, id DESC
    ) AS rn
  FROM notification
  WHERE user_id IS NULL
)
DELETE FROM notification n
USING ranked r
WHERE n.id = r.id
  AND r.rn > 1
RETURNING n.id, n.notif_type, n.additional_data;

COMMIT;
```

## How Has This Been Tested?

Added new tests. Would need to manually verify that this properly fixes it. 

## Additional Options

- [ ] [Optional] Please cherry-pick this PR to the latest release version.
- [x] [Optional] Override Linear Check

---
## Summary by cubic
Fixes anonymous-user migration failures by deduping and cleaning up NULL-owned `notification` rows and using savepoints with strict error handling to avoid partial migrations. Deletes rows that collide with existing anonymous-owned notifications and adds regression tests for duplicates and collisions.

- **Bug Fixes**
  - Deduplicate NULL-owned `notification` rows by `(notif_type, COALESCE(additional_data, '{}'::jsonb))`; delete any that would collide with existing anonymous-owned rows; run this dedup outside the savepoint so deletions persist even if the update rolls back.
  - Wrap each table’s update in a savepoint in upgrade and downgrade; stop swallowing errors and re-raise non-`notification` `IntegrityError`s to prevent partial migrations, while handling expected `notification` conflicts without aborting.

<sup>Written for commit d4e2ad04ce9836c372a0da0457192dc5d0fbb713. Summary will update on new commits.</sup>

### CODE REVIEW

**Comment by jessicasingh7** [comment]
looks good to me- would just check ai comments! @justin-tahara

**Comment by jessicasingh7** [comment]
@greptile

File: backend/alembic/versions/e7f8a9b0c1d2_create_anonymous_user.py [hunk 1]
```diff
@@ -65,7 +115,12 @@ def upgrade() -> None:
 
     # Migrate any remaining user_id=NULL records to anonymous user
     for table in TABLES_WITH_USER_ID:
-        try:
+        # Dedup notifications outside the savepoint so deletions persist
+        # even if the subsequent UPDATE rolls back
+        if table == "notification":
+            _dedupe_null_notifications(connection)
+
+        with connection.begin_nested():
```
[the remark this answers is not in the document]
Comment (jessicasingh7) [reply backend/alembic/versions/e7f8a9b0c1d2_create_anonymous_user.py:123]: The savepoint protects the notification dedup work from being rolled back if the UPDATE fails. We intentionally don't catch errors — if a table update fails, the migration should abort rather than silently succeed

File: backend/tests/integration/tests/migrations/test_migrations.py [hunk 2]
```diff
@@ -322,3 +322,165 @@ def test_jira_connector_migration() -> None:
...
+            ),
+            {"user_id": ANONYMOUS_USER_UUID},
+        ).fetchone()
+
+    assert len(notifications) == 1
+    assert notifications[0].id == 2  # Higher id wins when timestamps are equal
+    assert str(notifications[0].user_id) == ANONYMOUS_USER_UUID
+    assert anonymous_user is not None
+    assert anonymous_user.email == "anonymous@onyx.app"
+    assert anonymous_user.role == "LIMITED"
```
[the remark this answers is not in the document]
Comment (jessicasingh7) [reply backend/tests/integration/tests/migrations/test_migrations.py:411]: The downgrade is a simple SET user_id = NULL — there's not much logic to regress on.

File: backend/alembic/versions/e7f8a9b0c1d2_create_anonymous_user.py [hunk 3]
```diff
@@ -36,6 +36,56 @@
...
+            WHERE id IN (
+                SELECT id
+                FROM ranked_null_notifications
+                WHERE row_num > 1
+            )
+            """
+        )
+    )
+    if result.rowcount > 0:
+        print(f"Deleted {result.rowcount} duplicate NULL-owned notifications")
```
[the remark this answers is not in the document]
Comment (jessicasingh7) [reply backend/alembic/versions/e7f8a9b0c1d2_create_anonymous_user.py:67]: This is an acceptable edge case for a one-time data migration. The most recently shown row is the right default to keep.

File: backend/tests/integration/tests/migrations/test_migrations.py [hunk 4]
```diff
@@ -322,3 +322,164 @@ def test_jira_connector_migration() -> None:
...
+                """
+            ),
+            {"user_id": ANONYMOUS_USER_UUID},
+        ).fetchone()
+
+    assert len(notifications) == 1
+    assert str(notifications[0].user_id) == ANONYMOUS_USER_UUID
+    assert anonymous_user is not None
+    assert anonymous_user.email == "anonymous@onyx.app"
+    assert anonymous_user.role == "LIMITED"
```
[the remark this answers is not in the document]
Comment (jessicasingh7) [reply backend/tests/integration/tests/migrations/test_migrations.py:411]: already tested for this

File: backend/alembic/versions/e7f8a9b0c1d2_create_anonymous_user.py [hunk 5]
```diff
@@ -36,6 +36,50 @@
...
+            USING notification AS anonymous_owned
+            WHERE null_owned.user_id IS NULL
+              AND anonymous_owned.user_id = :user_id
+              AND null_owned.notif_type = anonymous_owned.notif_type
+              AND COALESCE(null_owned.additional_data, '{}'::jsonb) =
+                  COALESCE(anonymous_owned.additional_data, '{}'::jsonb)
+            """
+        ),
+        {"user_id": ANONYMOUS_USER_UUID},
+    )
```
[the remark this answers is not in the document]
Comment (jessicasingh7) [reply backend/alembic/versions/e7f8a9b0c1d2_create_anonymous_user.py:82]: stale- already added logging for this
